packages/hooks/src/useEventLogger.ts

```diff
@@ -0,0 +1,59 @@
+import * as React from 'react'
+import create from 'zustand'
+
+type EventType = string
+
+export interface Event {
+  name: EventType
+  payload: any
+}
+
+type EventReducer = (event: Event) => Event
+
+type EventLoggerStore = {
+  logEvent: (event: Event) => void
+  eventReducers: EventReducer[]
+  setEventReducers: (eventReducers: EventReducer[]) => void
+}
+
+// Middleware is just used to reduce/transform/process a single event
+// [transform1, transform2] => transform2(transform1(event)) : Event
+export const useEventLoggerStore = create<EventLoggerStore>(
+  (set, get): EventLoggerStore => ({
+    logEvent: async (event: Event) => {
+      get().eventReducers?.reduce((event, reducer) => reducer(event), event)
+    },
+    eventReducers: [],
+    setEventReducers: (eventReducers: EventReducer[]) =>
+      set({
+        eventReducers,
+      }),
+  }),
+)
+
+export function useEventLogger() {
+  const logEvent = useEventLoggerStore(store => store.logEvent)
+
+  return logEvent
+}
+
+const emptyArray = []
+
+export function useEventLoggerReducers(
+  { initialEventReducers },
+  deps?: React.DependencyList, // if we need to re-initialize the event reducers there needs to be a dependency array
+) {
+  const setEventReducers = useEventLoggerStore(store => store.setEventReducers)
+  const eventReducers = useEventLoggerStore(store => store.eventReducers)
```

Comment:
@alexanderchan we should memoize these selectors. You can move them outside of the hook or use `useCallback`.

https://github.com/react-spring/zustand#memoizing-selectors

Reply:
Oh good call!  I thought you'd had done that to make it easier to re-use the selectors but that's good to keep these functions the same for the re-renders.  I guess we can pull them outside.